Sandbox

ianoti opened the pull request:
I will fix my commit messages to improve the ease of following my workflow. I messed up on the commit messages. I also have yet to comment on much of the functions but I think the variable naming scheme makes it easier to follow. Please go through my code for the review. Thanks.

ianoti committed:
these are the local files pushed to git as a backup

ianoti committed:
(chore) initial application configuration setup with environments and models

ianoti committed:
(chore) added configuration file for build on CircleCI

ianoti committed:
(chore) added requirements.txt

ianoti committed:
(feature) added test command to manage.py

ianoti committed:
(bugfix) modified User models and blueprint in bucketlist

ianoti committed:
(chore) updated requirements

ianoti committed:
(tests) added basic tests for password hashing functionality and salting

ianoti committed:
(tests) added tests for user login functionality

ianoti committed:
(chore) random commits that shall be cleaned up later from sandbox

ianoti committed:
(chore) code backup on github

ianoti committed:
(chore) working dump of the code

ianoti committed:
(tests) refactored tests to use flask-testing

ianoti committed:
(chore) enable expiry time for token to be set from config file

ianoti committed:
(chore) unnecessary commits removed and added validation functions

ianoti committed:
(chore) refactored test to inherit from this module, set up the base test case

ianoti committed:
(chore) updated requirements.txt

dmigwi commented on README.md:
```diff
@@ -0,0 +1 @@
+python hello.py db upgrade to runinitial create database and tables
```
Ensure you have a proper Readme file.

dmigwi commented on app/authenticate.py:
```diff
@@ -0,0 +1,35 @@
+def verify_password(username, password):
+    user = User.query.filter_by(username=username).first()
+    if not user:
+        print("just a check")
```
`print("just a check")` should not be a part of you code, remove it.

dmigwi commented on app/authenticate.py:
```diff
@@ -0,0 +1,35 @@
+        return False
+    if not user.verify_password(password):
+        return False
+    g.user = user
```
Consider writing line 23-27 as: 
```
if not user or not user.verify_password(password):
        return False:
```

ianoti replied:
I modified the code to remove password based authentication since the application uses token based security.

dmigwi commented on app/models.py:
```diff
@@ -0,0 +1,80 @@
+            """ invalid token """
+            return False
+        user_id = data['id']
+        return user_id
```
* Avoid writing unnecessary code also only create the variables that you need.
you can replace the two exceptions that you are checking with just `except:`

Consider writing line of code 33-45 like as 
```
@staticmethod
     def token_confirm(token):
        s = Serializer(current_app.config['SECRET_KEY'])
        try:
              data = s.loads(token)
        except :
              """ invalid token """
              return False
        return data['id']
```
Also variable `user_id` isn't necessary you can just use `data['id']` coz its easier to read and interprate your code.

dmigwi commented on app/models.py:
```diff
@@ -0,0 +1,80 @@
+    status = db.Column(db.Boolean, default=False)
+    bucket_id = db.Column(db.Integer, db.ForeignKey('bucketlists.id'),
+                          nullable=False)
+
```
In your three models you have common fields. To avoid repeating fields like `id`, `name`, `date-created` e.t.c. You can create a class called base model that is inherited by all the other models.
Checkout what i did here.. https://github.com/andela-dmigwi/A-BucketList/blob/master/app/models.py if you need in understanding why the code  am available.

dmigwi commented on app/views.py:
```diff
@@ -0,0 +1,207 @@
+        username = args["username"]
+        password = args["password"]
+        email = args["email"]
+        if not validate_string(username):
```
for line 44-46 consider:
`username, password, email = args["username"], args["password"], args["email"]`

dmigwi commented on app/views.py:
```diff
@@ -0,0 +1,207 @@
+            return {"message": "no blank fields allowed"}, 400
+        # validate the email field
+        if not ("@" in email):
+            return {"message": "email is invalid"}, 400
```
that is not a good way of validating your email. Check out this site to help you write proper Regex expressions https://regex101.com/

dmigwi replied:
Consider using:
```
import re
result = re.match('\S+[@]\w+[.]\w+', 'email-bieng-tested')
//result should be true of false
```

dmigwi commented on migrations/README:
```diff
@@ -0,0 +1 @@
+Generic single-database configuration.
```
System generated files are not pushed to github.

dmigwi commented on migrations/alembic.ini:
```diff
@@ -0,0 +1,45 @@
+# A generic, single database configuration.
```
System generated files are not pushed to github.

kimani-ndegwa replied:
Just to correct @andela-dmigwi, migrations are intended to be pushed to github because they form part of the codebase. This link explains more:
http://stackoverflow.com/questions/28035119/should-i-be-adding-the-django-migration-files-in-the-gitignore-file

dmigwi commented on migrations/script.py.mako:
```diff
@@ -0,0 +1,24 @@
+"""${message}
```
System generated files are not pushed to github.

dmigwi commented on migrations/env.py:
```diff
@@ -0,0 +1,87 @@
+from __future__ import with_statement
```
System generated files are not pushed to github.

dmigwi commented on migrations/versions/48d121df1852_initial_migration.py:
```diff
@@ -0,0 +1,58 @@
+"""initial migration
```
System generated files are not pushed to github.

dmigwi reviewed (commented):
Please work on the issues that I have raised. Otherwise good work.

dmigwi commented:
Consider intergrating Travis ci too with your code.

ianoti committed:
(chore) removed debugging print statement

ianoti committed:
(test) added tests for pagination search and reordered file arrangement of tests

ianoti committed:
(chore) edited docstring for function that validates input as not empty

ianoti committed:
(chore) added error handling to app and pagination to the get requests for bucketlists

ianoti committed:
(chore) added custom 404 error handler to return json errors instead of html

ianoti committed:
(chore) added status badge to README document

ianoti committed:
(chore) minor changes to files
 1. added .travis.yml for travis integration
 2. implemented feedback from PR review

ianoti committed:
(chore) added script for tests

kimani-ndegwa commented on app/authenticate.py:
```diff
@@ -0,0 +1,34 @@
+from app.models import User
+
+token_auth = HTTPTokenAuth("Bearer")
+passw_auth = HTTPBasicAuth()
```
Consider using the full name for better understanding of your code base. Instead of `pass_w` use the full `password_authentication`. Its more friendly.

ianoti replied:
I decided to remove password based authentication, the application handles all security via tokens.

kimani-ndegwa commented on app/models.py:
```diff
@@ -0,0 +1,75 @@
+    id = db.Column(db.Integer, primary_key=True)
+    username = db.Column(db.String(255), unique=True, nullable=False)
+    email = db.Column(db.String(255), nullable=False)
+    pass_hash = db.Column(db.String(255), nullable=False)
```
Refer to my earlier comment on using full variable names.

kimani-ndegwa commented on app/validate_format.py:
```diff
@@ -0,0 +1,43 @@
+from . import db
+
+itemformat = {"id": fields.Integer,
+              "name": fields.String,
```
Consider separating this part into a possible `serializers.py` module. For separation of concerns

kimani-ndegwa commented on app/views.py:
```diff
@@ -0,0 +1,228 @@
+
+
+class ItemAction(Resource):
+    decorators = [multi_auth.login_required]
```
For purposes of convention, consider putting this decorator after the `__init__` function of your class.

kimani-ndegwa commented on config.py:
```diff
@@ -0,0 +1,46 @@
+    SQLALCHEMY_TRACK_MODIFICATIONS = False
+
+    @staticmethod
+    def init_app(app):
```
This function does not seem to have any implementation. Consider removing it.

ianoti replied:
I use the function in the application factor in the application factory in the create_app function in the __init__.py file in the app folder. It allows the global Config settings to be accessed by the app made in the factory, similar to how SQLAlchemy does it.

kimani-ndegwa commented on config.py:
```diff
@@ -0,0 +1,46 @@
+    "default": Development
+    }
+
+expiry_time = 3600
```
Consider putting this in the base `Config` class so that it does not look misplaced.

kimani-ndegwa commented on tests/test_users.py:
```diff
@@ -0,0 +1,169 @@
+        resp_nopassdata = json.loads(resp_nopass.data)
+        self.assertListEqual([400, 400], [resp_noname.status_code,
+                                          resp_nopass.status_code])
+        self.assertEqual(resp_nonamedata["message"]["username"], ("username"
```
Consider the naming convention suggested earlier.

kimani-ndegwa commented on tests/test_api_items.py:
```diff
@@ -0,0 +1,103 @@
+
+    def test_add_item(self):
+        """ test for successful addition of an item """
+        old = Item.query.count()
```
Why not use a variable like `old_count` instead of just old?

kimani-ndegwa commented on tests/test_api_items.py:
```diff
@@ -0,0 +1,103 @@
+        """ test that blank spaces, empty arguments not accepted when adding
+        item"""
+        blank = json.dumps({})
+        empty = json.dumps({"name": ""})
```
Good thinking here.

kimani-ndegwa commented on tests/test_api_items.py:
```diff
@@ -0,0 +1,103 @@
+    def test_update_item(self):
+        """ test that item can be updated and that at least
+        one argument must be provided"""
+        valid_data = json.dumps({"name": "updated list", "status": "true"})
```
This part could be abstracted using a base class where you define all the necessary properties and attributes and then inherit that class.  Peep this example and let me know in case something does not make sense :https://github.com/andela-kndegwa/CP3/blob/ch-write-tests/zuhura/api/tests/utils.py

ianoti committed:
(chore) added coveralls to travis.yml

ianoti committed:
(feature) added coverage badge

ianoti committed:
(chore)

ianoti committed:
(chore) added cover files to ignore list

ianoti committed:
(chore) removed relative import and increased test coverage

ianoti committed:
(chore) removed relative imports and unneccessary lines of code

ianoti committed:
(chore) edited test running scripts

ianoti committed:
(chore) split serialiser and utility functions

1. create separate file for serialiser
2. create separate file for utility functions

ianoti committed:
(chore) updated changed imports
